### PR TITLE
Add deprecation warnings when deprecated configs are set at compilation and execution

### LINKED ISSUE #17791: Deprecation and configs
While it seems pretty easy to add a warning when a deprecated config var, const or param is referenced in Chapel code, I think the behavior when it is referenced via the compilation or execution command is a little more up in the air so wanted to double check before implementing my instincts.

For example, say you have the following:
```chapel
deprecated "don't use x, use y" config var x: bool = false;
config var y: bool = false;

proc main() {
  writeln(x);
  writeln(y);
}
```
and you compile it with `chpl -sx=true foo.chpl`.  Should the act of using the compilation flag trigger a separate warning message from the writeln in the code?  What about if it was set at execution time, e.g. `./foo --x=true`?

My instinct says the flag use should trigger its own warning message.

### PR DESCRIPTION
**Pull request by lydia-duncan**
Deprecated configs already have warnings generated when the config is used in
Chapel code, but special handling was needed to generate warning when the
config is set via compilation or execution arguments.  This change adds such
handling.

Resolves #17791 
Resolves Cray/chapel-private#2089

<details>
Since we check whether a config has been set at compilation time when we see
the config's declaration (and don't attach the deprecation information until later),
generate the compilation time warning when we would attach the deprecation flag
and message.

For the execution flags, pass deprecation information along at codegen so that the
runtime can detect if the config being set was deprecated and generate the
appropriate message if so.
</details>

Adds tests to cover the behavior of config params and the behavior of config
vars, both with and without an explicit message provided.  These tests pass in
various combinations of compilation options and execution options respectively.

Passed a full paratest with futures

### CODE REVIEW

**Comment by lydia-duncan** [comment]
The compilation warning isn't quite as useful as the execution time warning, which indicates the flag that triggered the warning.  I'm thinking I might add a `USR_NOTE` to indicate that the symbol was set via a compilation flag to help distinguish it better from just regular uses in Chapel code?

**Comment by lydia-duncan** [comment]
It seems that this branch is sensitive to CHPL_DEVELOPER for some reason.  A bunch of tests with execution flags started printing out `<command-line arg>:1: warning: bool` with the change, but rebuilding with CHPL_DEVELOPER=true makes the error go away.

**Comment by lydia-duncan** [comment]
It also doesn't seem to fail on my Mac.  Maybe that's a C compiler issue then?

**Comment by lydia-duncan** [comment]
I don't know what's going on.  The generated C code looks right, but the call it is supposedly making has different values for the arguments and gdb won't let me look at the particularly frame in the backtrace that is making the call, even if I save the generated code using `--savec` (though, do we need a different strategy for debugging in a CHPL_LLVM=system world?).  I'm assuming the different arguments are what is causing the problem but I really don't see how it's my code's fault, maybe it's something to do with having touched the runtime to make the change?  I'm concerned but I don't know what to do about it

**Comment by lydia-duncan** [comment]
@bradcray - would you be up for reviewing this?

**Comment by bradcray** [comment]
Not this week, but maybe next.

**Comment by lydia-duncan** [comment]
If you haven't had a chance by EOD Tuesday, I'll ping someone else :)

File: test/deprecated-keyword/handleConfig.set.good [hunk 1]
```diff
@@ -0,0 +1,4 @@
+handleConfig.chpl:5: In function 'main':
+handleConfig.chpl:6: warning: don't use x, use y
+true
+<command-line arg>:1: warning: don't use x, use y
```
Comment (bradcray) [inline test/deprecated-keyword/handleConfig.set.good:4]: Is this the actual order that things print out, or is this reflecting some artifact of our test system w.r.t. stderr and stdout?  (I would've expected the command line warning to occur before `true` printed out).

File: test/deprecated-keyword/handleConfigParam.chpl [hunk 2]
```diff
@@ -0,0 +1,7 @@
+// Handle when a config param is referenced via its flag
+deprecated "don't use x, use y" config param x: bool = false;
```
Comment (bradcray) [inline test/deprecated-keyword/handleConfigParam.chpl:2]: `deprecated` can also not take a string, right?  How confident are you about the orthogonality of the code such that, if no string is used, the compile-time and execution-time warnings will also print as expected?

File: test/deprecated-keyword/handleConfigParam.set.good [hunk 3]
```diff
@@ -0,0 +1,5 @@
+warning: don't use x, use y
+note: 'x' was set via a compilation flag
```
Comment (bradcray) [inline test/deprecated-keyword/handleConfigParam.set.good:2]: Do we use this "compilation flag" term elsewhere?  My head goes to phrases like "compiler command-line option" or "compiler command-line flag" or "compiler option" or "compiler flag".  It looks like our man page uses "OPTIONS" when introducing them, but then "flags" when describing them.

File: compiler/codegen/codegen.cpp [hunk 4]
```diff
@@ -2052,6 +2055,9 @@ codegen_config() {
 
         args[3] = info->irBuilder->getInt32(var->hasFlag(FLAG_PRIVATE));
 
+        args[4] = info->irBuilder->getInt32(var->hasFlag(FLAG_DEPRECATED));
+        args[5] = info->irBuilder->CreateLoad(new_CStringSymbol(var->getDeprecationMsg())->codegen().val);
```
Comment (bradcray) [inline compiler/codegen/codegen.cpp:2059]: When you were getting LLVM errors last week, was it due to not having these lines and the ones just above?  Or something else?  (just curious how that resolved).

**Comment by bradcray** [comment]
Other than the comments above, this generally looked good to me.  I'm not familiar enough with the LLVM code generator to have high confidence in my ability to review the code.  For the flag, I pattern matched against the previous one and things looked fine.  For the string argument, I just assumed it was correct and read on.

File: test/deprecated-keyword/handleConfigParam.chpl (hunk 2 above)
Comment (lydia-duncan) [reply test/deprecated-keyword/handleConfigParam.chpl:2]: Very confident, but it's also not a burden to add a test in this case.

File: compiler/codegen/codegen.cpp (hunk 4 above)
Comment (lydia-duncan) [reply compiler/codegen/codegen.cpp:2059]: Yeah, it was due to not having these lines

File: test/deprecated-keyword/handleConfig.set.good (hunk 1 above)
Comment (lydia-duncan) [reply test/deprecated-keyword/handleConfig.set.good:4]: It's an artifact of our test system, running it manually printed the warning first

File: test/deprecated-keyword/handleConfigParam.chpl (hunk 2 above)
Comment (bradcray) [reply test/deprecated-keyword/handleConfigParam.chpl:2]: I don't feel strongly about it if you feel confident that it's orthogonal w.r.t. other existing tests.

**Pull request merged**